frontend/src/textToSpeech/textToSpeech.js

```diff
@@ -0,0 +1,383 @@
+        const { textData, sentenceIndex } = this.state;
+        const listLength = textData.length;
+        const newsentenceIndex = (sentenceIndex + delta + listLength) % listLength;
+        this.setState({
+            sentenceIndex: newsentenceIndex,
+            showNext: true,
+        });
+        this.reset();
+    }
+
+    reset = async () => {
+        this.setState({
+            showModal: false,
+            userText: '',
+            graded: false,
+            grade: null,
+            continue: false,
+            showAnswer: false,
+        });
+        document.getElementById('content').value = '';
+    }
+
+    giveUp = () => {
+        this.setState({
+            continue: true,
+            showAnswer: true,
+        });
+    }
+
+    playAudio = () => {
+        const utterance = new SpeechSynthesisUtterance();
+        utterance.text = this.getCurrentSentence();
+        utterance.lang = 'en-US';
+        utterance.rate = 1;
+        speechSynthesis.speak(utterance);
+    }
+
+    handleSubmit = () => {
+        const input = document.getElementById('content').value;
+        console.log('this is the input i want', input);
+        this.setState({
+            userText: input,
+        });
+        this.gradeText(input);
+
+        console.log('You have submitted!!!');
+    };
+
+    gradeText = async (input) => {
+        console.log('gradeText was called');
+        try {
+            console.log('im trying something here');
+            const apiURL = `/api/get_sentence_grade/${input}/${this.getCurrentSentence()}`;
+            const response = await fetch(apiURL);
+            const grade = await response.json();
+            console.log(grade);
+
+            this.setState({ grade: grade, graded: true });
+            this.setState({ continue: grade['isCorrect'] });
+        } catch (e) {
+            console.log(e);
+        }
+    }
+
+    modalHandler = (event) => {
+        event.preventDefault();
+        this.setState({
+            showModal: !this.state.showModal,
+        });
+    }
+
+    checkProgress = () => {
+        const numSentences = this.state.textData.length;
+        const currentSentence = this.state.sentenceIndex + 1;
+        if (numSentences !== 0) {
+            return parseInt((currentSentence / numSentences) * 100);
+        }
+        return 0;
+    }
+
+    giveGrade = () => {
+        const words = [];
+        for (let i = 0; i < this.state.grade['length']; i++) {
+            if (this.state.grade[i]['grade'] === 'correct') {
+                words.push(
+                    <span className="correct-word">{this.state.grade[i]['word']}</span>,
+                );
+                words.push(<span> </span>);
+            } else {
+                words.push(
+                    <span className="incorrect-word">{this.state.grade[i]['word']}</span>,
+                );
+                words.push(<span> </span>);
+            }
+        }
+        return words;
+    }
```

Comment:
This method stands out for a couple of reasons: 

1. There's more logic here than seems necessary for determining the string `className`. Could we simplify this with something like:
```
for (let i = 0; i < this.state.grade['length']; i++) {
  word = this.state.grade[i]['word']
  grade = this.state.grade[i]['grade']
  isWordCorrect = grade === 'correct';
  classNameString = isWordCorrect ? 'correct-word' : 'incorrect-word';
  words.push(<span className={classNameString}>{word}</span>);
  words.push(<span> </span>);
}
```

2. It seems odd to me to look to the top level of the `grade` object for a `length` property to determine the index to use to search other top-level properties of the `grade` object. Would it be possible to, instead, create a new property on `grade` that's an array of the `word` objects that you currently access by the top-level index (as strings)? That way the length of these graded words would be built in to the data structure, and you can simply iterate over the array.

Reply:
Thank you for both your suggestions. The formatting has been changed and the length key has been removed altogether as the length instance variable of the new array(which now holds all the words in their order) is now used instead.